common/lib/client/auth.js

```diff
@@ -145,45 +147,28 @@ var Auth = (function() {
 			callback = authOptions;
 			authOptions = null;
 		}
-
-		var self = this,
-			token = this.tokenDetails;
-
-		var requestToken = function() {
-			self.requestToken(tokenParams, authOptions, function(err, tokenResponse) {
-				if(err) {
-					callback(err);
-					return;
-				}
-				callback(null, (self.tokenDetails = tokenResponse));
-			});
+		callback = callback || noop;
+		var self = this;
+
+		/* _save normalises the tokenParams and authOptions and updates the auth
+		 * object. All subsequent operations should use the values on the object,
+		 * not the passed in ones.
+		 *
+		 * RSA10a: authorise() call implies token auth. If a key is passed it, we
+		 * just check if it doesn't clash and assume we're generating a token from it */
+		if(authOptions && authOptions.key && (this.key !== authOptions.key)) {
+			throw new ErrorInfo('Unable to update auth options with incompatible key', 40102, 401);
 		}
+		this._saveTokenOptions(tokenParams, authOptions);
 
-		if(token) {
-			if(this._tokenClientIdMismatch(token.clientId)) {
-				callback(new ErrorInfo('ClientId in token was ' + token.clientId + ', but library was instantiated with clientId ' + this.clientId, 40102, 401));
-				return;
-			}
-			this.getTimestamp(self.authOptions && self.authOptions.queryTime, function(err, time) {
-				if(err)
-					callback(err);
+		this._logTokenAuthMethod();
 
-				if(token.expires === undefined || (token.expires >= time)) {
-					if(!(authOptions && authOptions.force)) {
-						Logger.logAction(Logger.LOG_MINOR, 'Auth.getToken()', 'using cached token; expires = ' + token.expires);
-						callback(null, token);
-						return;
-					}
-				} else {
-					/* expired, so remove */
-					Logger.logAction(Logger.LOG_MINOR, 'Auth.getToken()', 'deleting expired token');
-					self.tokenDetails = null;
-				}
-				requestToken();
-			});
-		} else {
-			requestToken();
-		}
+		this._ensureValidAuthCredentials(function() {
```

Comment:
Fair enough. (also allows for a !err check before disconnecting, which is probably a good idea)